### ISSUE TITLE
RFC: Internal Unification of LRScheduler Semantics for Correct Resuming and Composition

### ISSUE THREAD

**Issue by diputs-sudo** [pytorch/pytorch#168044]
### 🚀 The feature, motivation and pitch

# RFC: Improving the Internal Consistency of PyTorch LR Schedulers (Inspired by Issue #162591)

## Summary
While trying to understand and fix the behavior described in **issue #162591**, where `SequentialLR` resumes training with completely unexpected learning rates, I ended up going much deeper into PyTorch's LR scheduler internals than I originally planned. What started out as a small bug investigation quickly turned into a discovery that many LR schedulers behave inconsistently, especially when resuming from a checkpoint.

After spending time reading the code, testing different schedulers, and comparing LR sequences across resumes, I realized these issues aren’t coming from one broken class, they’re coming from assumptions inside the whole scheduler system. This RFC is my attempt to summarize what I found and propose a way to clean things up internally while keeping the existing public API unchanged.

I’m doing this because I want PyTorch’s schedulers to be predictable and reliable, and because digging into this part of the library has been genuinely interesting.

## Motivation
While debugging `SequentialLR`, I repeatedly hit situations where:
- resuming from `last_epoch=N` gave LR values that didn't match the values produced in a continuous run,
- milestone transitions behaved differently depending on how the schedulers were chained,
- some schedulers silently performed a step during initialization,
- stateful schedulers didn’t store enough internal state to reproduce their behavior after loading a checkpoint,
- closed form and chainable schedulers interacted unpredictably.

It became clear that the schedulers weren’t built on a consistent internal contract. They each have their own interpretation of concepts like `last_epoch`, how initial LR is handled, and how state should be restored.

I’m proposing internal changes not because the API is flawed. It’s actually quite flexible, but because the internal logic needs to be aligned for the behavior to be correct.

## Goals
- Make resuming training reliable across **all** LR schedulers.
- Ensure `SequentialLR` and `ChainedScheduler` behave consistently when switching between schedulers.
- Unify internal behavior, initialization, `last_epoch`, closed form logic, and state handling, without changing how users interact with the API.
- Guarantee that LR sequences match exactly between resumed and uninterrupted runs.
- Reduce surprising or hidden behavior inside the scheduler system.

## Root Causes (As Observed During Investigation)
1. **Different schedulers interpret `last_epoch` differently**, leading to mismatched LR updates.
2. **Implicit initialization steps** occur in some schedulers, causing LR to shift before the user even calls `step()`.
3. **A mix of chainable and closed form logic** exists, but SequentialLR assumes all schedulers behave uniformly.
4. **Stateful schedulers don’t fully serialize their internal state**, making them impossible to reconstruct exactly.
5. **Scheduler composition relies on assumptions** that hold for some schedulers but break others.

### 1. Make `last_epoch` mean the same thing everywhere
Every scheduler should treat `last_epoch` as the number of LR updates that have been completed.
This alone eliminates many inconsistencies.

### 2. Remove implicit LR updates during initialization
Schedulers should not alter the LR before the user's first `step()` call.

### 3. Provide internal closed form LR computation
Schedulers can implement:
```python
def _compute_lr_closed_form(self, epoch: int, group_index: int): ...
```
This ensures predictable LR values during resume, without affecting the public API.

If a scheduler cannot compute closed form LR (like CyclicLR), it should provide a deterministic fast forward mode that **doesn’t** update optimizer state.

### 4. Ensure `state_dict` stores everything needed for full reconstruction
This includes cycle counters, restart phases, internal LR or momentum transitions. Anything that affects scheduler behavior.

### 5. Rewrite `SequentialLR` to rely on the unified internal logic
Instead of calling child schedulers’ `step()` methods or trying to rewind them, `SequentialLR` should:
- calculate the correct local epoch,
- use closed form/fast forward logic to get LR,
- avoid modifying child scheduler state during composition.

### 6. Guarantee reproducibility across resumes
If a run reaches step 100, saving and resuming at `last_epoch=100` should produce **exactly** the same LR sequence as if the training never stopped.

## Incremental Rollout Plan
I’m happy to work on this and contribute as much as I can. I also want to do it in a steady and manageable way so it is easy for maintainers to follow and review.

### Step 1
Start with the simpler schedulers and make their behavior consistent around initialization and `last_epoch`. This lets me establish the foundation without touching the more complex parts yet.

### Step 2
Add internal closed form helpers where they make sense so schedulers can reliably tell what LR they should produce at any step.

### Step 3
Work on the more stateful schedulers like CyclicLR, OneCycleLR, and CosineAnnealingWarmRestarts. The goal here is to make their internal state easy to restore and to support deterministic fast forward behavior.

### Step 4
Once the individual schedulers are consistent, refactor `SequentialLR` to rely on this unified logic so it behaves predictably when switching between schedulers.

### Step 5
Wrap up by cleaning old behavior that is no longer needed and writing tests to make sure everything stays consistent.

## Backward Compatibility
Everything here happens inside the implementation. No public API changes are required:
- No argument changes.
- No new user methods.
- No behavior changes in how schedulers are called.

The goal is purely internal correctness and consistency.

## Conclusion
Working on the `SequentialLR` bug made it clear that the real issue wasn’t that one scheduler, but the assumptions the whole system is built on. This RFC proposes a way to align that internal logic without breaking user code.

I’m sharing this with the aim of making scheduler behavior more predictable and easier to reason about. Something I believe will help many users who resume training or combine schedulers.

I’d appreciate any feedback, direction, or suggested adjustments. I’m happy to work on this incrementally and collaborate along the way.

Thank you for taking the time to read through this!

cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar

**Comment by diputs-sudo** [comment]
Quick update: I’ve opened the first incremental PR related to this RFC:

https://github.com/pytorch/pytorch/pull/168223

This PR adds a failing test that captures one of the issues mentioned here(negative epoch values leading to a ZeroDivisionError in closed form LR schedulers). It’s test only as discussed, and intended to document current behavior before applying internal fixes.

I’ll continue adding failing tests for the remaining resume and consistency problems before proposing any internal unification changes.

**Comment by diputs-sudo** [comment]
Quick follow up: I opened another failing test PR related to one of the inconsistencies I came across while checking closed form vs recursive LR behavior.

PR #169107: PolynomialLR `total_iters=0` closed form crash

This test exposes two related issues in `PolynomialLR(total_iters=0)`:

- Closed form stepping (`step(epoch=...)`) produces a `0 / 0` division, which crashes
- Recursive mode returns base_lr forever instead of the expected end_lr

The purpose of the PR is just to document the current behavior before unifying how schedulers handle edge cases like `total_iters=0`.

I’ll keep adding these targeted failing tests for each scheduler so the larger cleanup can be done with full coverage.

**Comment by diputs-sudo** [comment]
Quick follow up: I opened another failing test PR related to scheduler initialization behavior that came up while investigating resume and composition semantics.

PR #171686: Implicit step during `LRScheduler` initialization

This test documents the current behavior where constructing an `LRScheduler` performs an implicit learning rate step during `__init__`, before any user visible call to `scheduler.step()`.

In particular:

* `LRScheduler.__init__()` unconditionally calls `_initial_step()`
* `_initial_step()` internally invokes `step()` once
* As a result, `last_epoch` is set to `0` and `_step_count` is incremented immediately after construction
* This happens even though the user has not yet called `scheduler.step()`

This implicit step is non obvious from the public API and contributes to off by one behavior, surprising resume semantics, and the need for compensating logic in composed schedulers such as `SequentialLR`.

The purpose of this PR is **tests only**. It documents the existing behavior and makes the implicit initialization step explicit in the test suite. A follow up change can revisit initialization semantics once there is consensus on the desired behavior.

I’ll continue adding small, targeted failing tests like this as I encounter them, so that any larger scheduler cleanup can proceed with clear coverage and fewer surprises.

**Comment by diputs-sudo** [comment]
Quick follow up: while continuing to investigate scheduler edge cases related to validation, resume behavior, and closed-form stepping, I opened another small tests-only PR documenting an additional failure mode in `CosineAnnealingLR`.

PR #174025: Add test for CosineAnnealingLR when T_max = 0

This test documents the current behavior of `CosineAnnealingLR` when constructed with `T_max = 0`. Internally, `CosineAnnealingLR` relies on `T_max` as both a divisor and a modulo base when computing the learning rate during `scheduler.step()`, but the constructor does not validate that `T_max > 0`.

As a result, calling `scheduler.step()` with `T_max = 0` currently raises a `ZeroDivisionError`. The failure only surfaces at step time and comes from low level math rather than an explicit validation check, which makes the behavior non obvious from the public API.

This PR is **tests only**. It does not propose a fix or define canonical semantics for `T_max = 0`. Instead, it makes the existing behavior explicit in the test suite so that future changes to validation or scheduler semantics can be made with clear coverage.

I’ll continue adding small, targeted tests like this as additional edge cases come up, so that any larger scheduler cleanup or refactor can proceed incrementally, with concrete coverage and fewer surprises.

**Comment by diputs-sudo** [comment]
Quick follow up: while continuing to investigate resume behavior and scheduler composition semantics, I opened another small tests-only PR documenting an additional inconsistency in `SequentialLR`.

PR #175506: Add failing test for SequentialLR resume reproducibility

This test documents the current behavior where `SequentialLR` does not reproduce the same learning rate sequence after saving and restoring optimizer and scheduler state.

From a user perspective, resuming training from a checkpoint should yield the exact same LR sequence as if the run had never been interrupted. However, when `SequentialLR` switches between schedulers at milestone boundaries, resuming from an intermediate step can produce a different LR sequence compared to an uninterrupted run.

The purpose of this PR is tests only. It does not propose a fix or redefine resume semantics. Instead, it makes the current behavior explicit in the test suite so that any future changes to scheduler state handling or composition logic can be made with clear coverage.

I’ll continue adding small, targeted tests like this where inconsistencies show up, so that any larger scheduler cleanup can proceed incrementally and with concrete guarantees around behavior.

**Comment by diputs-sudo** [comment]
Quick follow up: I opened another failing test PR related to `OneCycleLR` edge cases that came up while looking into scheduler phase construction.

PR #180187 : Degenerate phase handling in `OneCycleLR` initialization

This test documents the current behavior where certain small schedules lead to degenerate phases during `OneCycleLR` construction.

In particular:

* For very small `total_steps`, some `pct_start` values produce phases with zero width (e.g. `total_steps=2, pct_start=0.5` or `total_steps=1, pct_start=1.0`)
* The scheduler still proceeds with phase construction instead of short-circuiting or validating inputs
* This leads to a division by zero in the phase math during `__init__`
* As a result, a `ZeroDivisionError` is raised at construction time
* Enabling `three_phase=True` does not avoid the issue for these tiny schedules

This behavior is not very obvious from the API surface and suggests that the implementation assumes strictly positive phase lengths without enforcing it.

The purpose of this PR is **tests only**. It documents the current behavior and makes these degenerate cases explicit in the test suite. A follow up change can either introduce proper validation (e.g. raising a `ValueError`) or define consistent handling for zero-length phases.

I’ll continue adding small, targeted failing tests like this as I run into them, so that any broader scheduler cleanup has clear coverage of these edge cases.

**Comment by diputs-sudo** [comment]
Quick follow up: while continuing to investigate scheduler initialization behavior and edge case handling, I opened another small tests only PR documenting additional initialization inconsistencies in `OneCycleLR`.

PR #183386 : Add tests for OneCycleLR zero divisor initialization behavior

These tests document the current behavior when `div_factor=0.0` or `final_div_factor=0.0`.

Internally, `OneCycleLR` uses these values during initialization to compute derived learning rates:

* `initial_lr = max_lr / div_factor`
* `min_lr = initial_lr / final_div_factor`

Because these values are currently not validated explicitly, zero values fall through to raw `ZeroDivisionError`s during scheduler construction.

The behavior is not very obvious from the API surface and suggests that `OneCycleLR` assumes both divisor values are strictly non-zero without enforcing that assumption directly.

This PR is tests only. It documents the current behavior and makes these initialization failure modes explicit in the test suite. A follow up change can introduce explicit validation or define consistent handling for these invalid configurations.

**Comment by diputs-sudo** [comment]
Quick follow up: while continuing to investigate scheduler validation and edge case handling, I opened another small tests only PR documenting an additional invalid configuration in ConstantLR.

PR #186568: Add expected-failure test for ConstantLR factor=0

This test documents the current behavior when `factor=0` is passed to `ConstantLR`.

Internally, `ConstantLR` uses `factor` to scale the learning rate during the constant phase and later restores the original learning rate by applying the inverse scaling. When `factor=0`, the scheduler can be constructed successfully, but a later step reaches a code path that computes:

`1 / factor`

which results in a raw `ZeroDivisionError`.

Because this value is not currently validated during initialization, the invalid configuration is accepted up front and only fails later during scheduler execution.

The behavior is not immediately obvious from the API surface and suggests that `ConstantLR` assumes `factor` is non zero without enforcing that assumption directly.

This PR is tests only. It documents the current behavior and makes the failure mode explicit in the test suite. A follow up change can introduce explicit validation or define consistent handling for this invalid configuration, such as raising a clearer `ValueError` during scheduler construction.
